Add complex IValues

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50885 Add serialization logic for complex numbers
* #50884 Add type annotation logic for complex numbers
* **#50883 Add complex IValues**

Differential Revision: [D26003682](https://our.internmc.facebook.com/intern/diff/D26003682)

anjali411 committed:
Add complex IValues

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex IValues"




[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1105,6 +1105,7 @@ using NumberTypePtr = std::shared_ptr<NumberType>;
 // Subtype hierarchy for Number Types (NumberType as the base type):
 // IntType <: NumberType
 // FloatType <: NumberType
+// ComplexDoubleType <:NumberType
```
I know I suggested this initially, but I discussed this with the JIT team and we concluded that it would be better to make `ComplexDoubleType` a subtype of only `AnyType`. `NumberType` is an internal trick used to make sure operators written in Python that take both `int` and `float` (e.g. `norm`, that takes an `int` or `float` for the type of norm to compute) are JITtable. `ComplexDoubleType` cannot be passed in for every parameter that is annotated `NumberType` today.

SplitInfinity replied:
As discussed offline, both actually make sense. Let's follow up on this and fix it later.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1156,6 +1157,33 @@ struct TORCH_API FloatType : public NumberType {
+    return "complex";
+  }
+  bool isSubtypeOfExt(const TypePtr& rhs, std::ostream* why_not) const override {
+    return rhs->kind() == TypeKind::NumberType || NumberType::isSubtypeOfExt(rhs, why_not);
```
Same here - `ComplexDoubleType` should be a subtype of only `Any`.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1525,6 +1553,8 @@ inline TypePtr TensorType::fromNumberType(TypePtr typ) {
     return TensorType::createContiguous(at::kFloat, at::kCPU, {});
   } else if (typ->isSubtypeOf(BoolType::get())) {
     return TensorType::createContiguous(at::kLong, at::kCPU, {});
+  } else if (typ->isSubtypeOf(ComplexDoubleType::get())) {
+    return TensorType::createContiguous(at::kComplexDouble, at::kCPU, {});
```
I don't think this change is necessary for this PR. I don't think it is being tested either.

anjali411 replied:
makes sense removed

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1535,6 +1565,8 @@ inline TypePtr TensorType::fromBoolType() {
 inline c10::optional<c10::ScalarType> tryScalarTypeFromJitType(const c10::TypePtr & type) {
   if (type == FloatType::get()) {
     return at::typeMetaToScalarType(c10::get_default_dtype());
+  } else if (type == ComplexDoubleType::get()) {
+    return at::typeMetaToScalarType(c10::get_default_complex_dtype());
```
I don't think this change is necessary for this PR. I don't think it is being tested either.

anjali411 replied:
makes sense removed

SplitInfinity reviewed (commented):
Nice! After further discussion with the JIT team about the place of `ComplexDoubleType` in the type system, we concluded that it makes more sense for it to be a subtype of only `Any`, and not of `NumberType`. Other than that, the rest looks good!

anjali411 committed:
Update on "Add complex IValues"




[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/test/type_test.cpp:
```diff
@@ -175,8 +175,8 @@ class OneForward(Interface):
   // Tuples should be structurally typed
-  auto type = TupleType::create({IntType::get(), TensorType::get(), FloatType::get()});
-  auto type2 = TupleType::create({IntType::get(), TensorType::get(), FloatType::get()});
+  auto type = TupleType::create({IntType::get(), TensorType::get(), FloatType::get(), ComplexDoubleType::get()});
```
Are these changes for testing `ComplexDoubleType::operator==`?

anjali411 replied:
yeah and also indirectly testing `ComplexDoubleType::get()`

SplitInfinity reviewed (approved):
Let's make sure to follow up on the correct place for `ComplexDoubleType` in the JIT type system and fix it if needed in the future.

Approval contingent on tests passing.

anjali411 commented:
> Nice! After further discussion with the JIT team about the place of `ComplexDoubleType` in the type system, we concluded that it makes more sense for it to be a subtype of only `Any`, and not of `NumberType`. Other than that, the rest looks good!

Synced offline: It probably makes sense for `ComplexDoubleType` to be a subtype of `NumberType` since we can have complex scalars and functions like `torch.full`, `torch.index_fill` etc. can take complex numbers as input.

anjali411 committed:
Update on "Add complex IValues"


Differential Revision: [D26003682](https://our.internmc.facebook.com/intern/diff/D26003682)

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex IValues"


Differential Revision: [D26003682](https://our.internmc.facebook.com/intern/diff/D26003682)

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex IValues"


Differential Revision: [D26003682](https://our.internmc.facebook.com/intern/diff/D26003682)

[ghstack-poisoned]
